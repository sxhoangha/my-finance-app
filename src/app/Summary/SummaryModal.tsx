import React from "react";
import Box from "@mui/material/Box";
import Button from "@mui/material/Button";
import Typography from "@mui/material/Typography";
import Modal from "@mui/material/Modal";
import { DataGrid, GridRowsProp, GridColDef } from "@mui/x-data-grid";
import { ITransaction } from "@/types";
import { formatDate } from "../utils";

const style = {
  position: "absolute" as "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  width: "45rem",
  bgcolor: "background.paper",
  border: "2px solid #000",
  boxShadow: 24,
  p: 4,
};

interface SummaryModalProps {
  transactions: Array<ITransaction>;
  open: boolean;
  handleClose: () => void;
  handleOpen: () => void;
}

const SummaryModal: React.FC<SummaryModalProps> = (props) => {
  const { transactions, open, handleClose } = props;

  const rows: GridRowsProp = transactions.map((transaction) => ({
    ...transaction,
    transactionDate: formatDate(transaction.transactionDate),
  }));

  const columns: GridColDef[] = [
    { field: "referenceNumber", headerName: "Reference Number", width: 150 },
    { field: "amount", headerName: "Amount", width: 150 },
    { field: "description", headerName: "Description", width: 150 },
    { field: "transactionDate", headerName: "Date", width: 150 },
  ];

  return (
    <div>
      <Modal
        open={open}
        onClose={handleClose}
        aria-labelledby="modal-modal-title"
        aria-describedby="modal-modal-description"
      >
        <Box sx={style}>
          <Typography id="modal-modal-title" variant="h6" component="h2">
            Your transactions
          </Typography>
          <DataGrid
            rows={rows}
            columns={columns}
            initialState={{
              pagination: { paginationModel: { pageSize: 10 } },
            }}
            pageSizeOptions={[5, 10]}
          />
          <Button onClick={handleClose}>Close</Button>
        </Box>
      </Modal>
    </div>
  );
};

export default SummaryModal;
