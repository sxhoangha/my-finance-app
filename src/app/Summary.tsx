"use client";
import Button from "@mui/material/Button";
import TextField from "@mui/material/TextField";
import { ITransaction } from "@/types";
import { BalanceColor } from "@/types/constants";
import React, { useState } from "react";
import SummaryModal from "./SummaryModal";

interface SummaryProps {
  transactions: Array<ITransaction>;
}

const Summary = ({ transactions }: SummaryProps) => {
  const [threshold, setThreshold] = useState(0);

  const [openModal, setOpenModal] = useState(false);
  const handleOpen = () => setOpenModal(true);
  const handleClose = () => setOpenModal(false);

  const balance = transactions.reduce((acc, transaction) => {
    return acc + transaction.amount;
  }, 0);

  const color =
    balance < 0
      ? BalanceColor.RED
      : balance < (isNaN(threshold) ? 0 : threshold)
      ? BalanceColor.YELLOW
      : BalanceColor.GREEN;

  return (
    <div>
      <h2>Summary</h2>
      <p>Here is a summary of your account</p>
      <div>
        Account balance:{" "}
        <span style={{ color: color }}>
          {balance} <Button onClick={handleOpen}>View transactions</Button>
        </span>
      </div>
      <TextField
        id="threshold"
        label="Threshold"
        type="number"
        variant="filled"
        InputLabelProps={{
          shrink: true,
        }}
        value={threshold}
        onChange={(e) => setThreshold(parseInt(e.target.value))}
        error={threshold < 0}
        helperText={threshold < 0 ? "Threshold cannot be negative" : ""}
      />

      <SummaryModal
        open={openModal}
        handleOpen={handleOpen}
        handleClose={handleClose}
        transactions={transactions}
      />
    </div>
  );
};

export default Summary;
