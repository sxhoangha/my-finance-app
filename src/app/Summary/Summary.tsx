"use client";
import { IInvoice, ITransaction } from "@/types";
import ReceiptLongIcon from "@mui/icons-material/ReceiptLong";
import IconButton from "@mui/material/IconButton";
import TextField from "@mui/material/TextField";
import Tooltip from "@mui/material/Tooltip";
import dayjs from "dayjs";
import React, { useState } from "react";
import { BalanceColor } from "../../types/constants";
import SummaryModal from "./SummaryModal";

interface SummaryProps {
  transactions: Array<ITransaction>;
  invoices: Array<IInvoice>;
}

const Summary = ({ transactions, invoices }: SummaryProps) => {
  const [threshold, setThreshold] = useState(0);

  const [openModal, setOpenModal] = useState(false);
  const handleOpen = () => setOpenModal(true);
  const handleClose = () => setOpenModal(false);

  const balance = transactions.reduce((acc, transaction) => {
    return acc + transaction.amount;
  }, 0);

  const color =
    balance < 0
      ? BalanceColor.RED
      : balance < (isNaN(threshold) ? 0 : threshold)
      ? BalanceColor.YELLOW
      : BalanceColor.GREEN;

  return (
    <div className="fi-summary">
      <h4 className="fi-summary-title">Summary</h4>
      <TextField
        id="threshold"
        label="Threshold"
        type="number"
        variant="filled"
        InputLabelProps={{
          shrink: true,
        }}
        value={threshold}
        onChange={(e) => setThreshold(parseInt(e.target.value))}
        error={threshold < 0}
        helperText={threshold < 0 ? "Threshold cannot be negative" : ""}
      />
      <div>
        Account balance: <span style={{ color: color }}>{balance}</span>
      </div>
      <div>
        Total number of transactions: {transactions.length}{" "}
        <Tooltip title="Transaction details">
          <IconButton onClick={handleOpen}>
            <ReceiptLongIcon />
          </IconButton>
        </Tooltip>
      </div>

      <div>
        Total number of invoices in the last 30 days:{" "}
        <span data-testid="number-of-invoices">
          {
            invoices.filter(({ creationDate }) =>
              dayjs(creationDate).isAfter(dayjs(new Date()).subtract(30, "day"))
            ).length
          }
        </span>
      </div>

      <SummaryModal
        open={openModal}
        handleOpen={handleOpen}
        handleClose={handleClose}
        transactions={transactions}
      />
    </div>
  );
};

export default Summary;
