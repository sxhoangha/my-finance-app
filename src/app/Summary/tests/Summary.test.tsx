import renderer from "react-test-renderer";
import React from "react";
import { render, screen } from "@testing-library/react";
import Summary from "../Summary";

describe("Summary", () => {
  const transactions = [
    {
      id: "1a3e4b5c-6d7e-8f9a-bcde-0123456789ab",
      transactionDate: new Date("2022-01-02T00:00:00Z"),
      description: "Salary",
      referenceNumber: "TRX001",
      amount: 5000,
    },
    {
      id: "2b4d6f8a-0c2e-4g6i-8k0m-123456789abc",
      transactionDate: new Date("2022-01-02T00:00:00Z"),
      description: "Rent",
      referenceNumber: "TRX002",
      amount: -1000,
    },
  ];

  const invoices = [
    {
      clientName: "hoang",
      creationDate: new Date("2021-11-01T14:20:04.961Z"),
      referenceNumber: "TRX001",
      amount: 5000,
      id: "g_LOYCFKgXH9M2PTyF5u9",
      status: "PAID",
    },
    {
      id: "64uj94Bgpo0qffLtzTxmD",
      clientName: "111",
      creationDate: new Date("2024-01-05T03:22:31.449Z"),
      referenceNumber: "111",
      amount: 11,
      status: "NOT PAID",
    },
  ];
  test("renders Summary component correctly", () => {
    const tree = renderer
      .create(<Summary transactions={transactions} invoices={invoices} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("number of invoice should be 1", () => {
    const { getByText } = render(
      <Summary transactions={transactions} invoices={invoices} />
    );
    const element = screen.getByTestId("number-of-invoices");
    expect(element.textContent).toBe("1");
  });
});
