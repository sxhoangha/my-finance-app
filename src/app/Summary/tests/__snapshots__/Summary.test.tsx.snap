// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary renders Summary component correctly 1`] = `
<div
  className="fi-summary"
>
  <h4
    className="fi-summary-title"
  >
    Summary
  </h4>
  <div
    className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-filled MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-filled css-o943dk-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor="threshold"
      id="threshold-label"
    >
      Threshold
    </label>
    <div
      className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-batk84-MuiInputBase-root-MuiFilledInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiFilledInput-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
        disabled={false}
        id="threshold"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="number"
        value={0}
      />
    </div>
  </div>
  <div>
    Account balance: 
    <span
      style={
        {
          "color": "green",
        }
      }
    >
      4000
    </span>
  </div>
  <div>
    Total number of transactions: 
    2
     
    <button
      aria-label="Transaction details"
      aria-labelledby={null}
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      data-mui-internal-clone-element={true}
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ReceiptLongIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19.5 3.5 18 2l-1.5 1.5L15 2l-1.5 1.5L12 2l-1.5 1.5L9 2 7.5 3.5 6 2v14H3v3c0 1.66 1.34 3 3 3h12c1.66 0 3-1.34 3-3V2l-1.5 1.5zM19 19c0 .55-.45 1-1 1s-1-.45-1-1v-3H8V5h11v14z"
        />
        <path
          d="M9 7h6v2H9zm7 0h2v2h-2zm-7 3h6v2H9zm7 0h2v2h-2z"
        />
      </svg>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div>
    Total number of invoices in the last 30 days:
     
    <span
      data-testid="number-of-invoices"
    >
      1
    </span>
  </div>
  <div />
</div>
`;
